velox/type/TypeCoercer.cpp

```diff
@@ -55,48 +56,54 @@ allowedCoercions() {
 } // namespace
 
 // static
-std::optional<Coercion> TypeCoercer::coerceTypeBase(
+Coercion TypeCoercer::coerceTypeBase(
     const TypePtr& fromType,
     const std::string& toTypeName) {
   static const auto kAllowedCoercions = allowedCoercions();
   if (fromType->name() == toTypeName) {
-    return Coercion{.type = fromType, .cost = 0};
+    return Coercion{fromType, 0};
+  }
+
+  if (fromType == UNKNOWN()) {
+    // Cast Unknown to complex type is not supported yet
+    return Coercion{getType(toTypeName, {}), kMinCoercionCost};
   }
 
   auto it = kAllowedCoercions.find({fromType->name(), toTypeName});
   if (it != kAllowedCoercions.end()) {
     return it->second;
   }
 
-  return std::nullopt;
+  return {};
 }
 
 // static
-bool TypeCoercer::coercible(const TypePtr& fromType, const TypePtr& toType) {
-  if (fromType->isUnKnown()) {
-    return true;
+Coercion TypeCoercer::coercible(
+    const TypePtr& fromType,
+    const TypePtr& toType) {
```

Comment:
this code can be called for same type (see If/SwitchExpr)
I think it's better to write
if (fromType->equal
```suggestion
Coercion TypeCoercer::coercible(
    const TypePtr& fromType,
    const TypePtr& toType) {
   if (fromType->equivalent(*toType)) {
      return {fromType, 0};
   }
```